controllers/awsmachinetemplate_controller.go

```diff
@@ -18,44 +18,124 @@ package controllers
-// the AWSMachineTemplate object against the actual cluster state, and then
-// perform operations to make the cluster state reflect the state specified by
-// the user.
-//
 // For more details, check Reconcile and its Result here:
 // - https://pkg.go.dev/sigs.k8s.io/controller-runtime@v0.8.3/pkg/reconcile
-func (r *AWSMachineTemplateReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
-	_ = log.FromContext(ctx)
+func (r *AWSMachineTemplateReconciler) Reconcile(req ctrl.Request) (ctrl.Result, error) {
+	var err error
+	ctx := context.TODO()
+	logger := r.Log.WithValues("namespace", req.Namespace, "awsMachineTemplate", req.Name)
+
+	awsMachineTemplate := &capa.AWSMachineTemplate{}
+	if err := r.Get(ctx, req.NamespacedName, awsMachineTemplate); err != nil {
+		logger.Error(err, "AWSMachineTemplate does not exist")
+		return ctrl.Result{}, err
+	}
+
+	clusterName := key.GetClusterIDFromLabels(awsMachineTemplate)
+
+	logger = logger.WithValues("cluster", clusterName)
+
+	var awsClientGetter *awsclient.AwsClient
+	{
+		c := awsclient.AWSClientConfig{
+			ClusterName: clusterName,
+			CtrlClient:  r.Client,
+			Log:         logger,
+		}
+		awsClientGetter, err = awsclient.New(c)
+		if err != nil {
+			logger.Error(err, "failed to generate awsClientGetter")
+			return ctrl.Result{}, err
+		}
+	}
+
+	awsClientSession, err := awsClientGetter.GetAWSClientSession(ctx)
+	if err != nil {
+		logger.Error(err, "Failed to get aws client session")
+		return ctrl.Result{}, err
+	}
+
+	var iamService *iam.IAMService
+	{
+		c := iam.IAMServiceConfig{
+			AWSSession:  awsClientSession,
+			ClusterName: clusterName,
+			IAMRoleName: awsMachineTemplate.Spec.Template.Spec.IAMInstanceProfile,
+			Log:         logger,
+			RoleType:    iam.ControlPlaneRole,
+		}
+		iamService, err = iam.New(c)
+		if err != nil {
+			logger.Error(err, "Failed to generate IAM service")
+			return ctrl.Result{}, err
+		}
+	}
+
+	if awsMachineTemplate.DeletionTimestamp != nil {
+		err = iamService.Delete()
+		if err != nil {
+			logger.Error(err, "failed to delete IAM Role")
+			return ctrl.Result{}, err
+		}
+		controllerutil.RemoveFinalizer(awsMachineTemplate, key.CAPAIAMControllerFinalizer)
+		err = r.Update(ctx, awsMachineTemplate)
+		if err != nil {
+			logger.Error(err, "failed to remove finalizer from AWSMachineTemplate")
+			return ctrl.Result{}, err
+		}
+
+	} else {
+		err = iamService.Reconcile()
+		if err != nil {
+			logger.Error(err, "failed to reconcile IAM Role")
+			return ctrl.Result{}, err
+		}
+
+		controllerutil.AddFinalizer(awsMachineTemplate, key.CAPAIAMControllerFinalizer)
+		err = r.Update(ctx, awsMachineTemplate)
+		if err != nil {
+			logger.Error(err, "failed to add finalizer on AWSMachineTemplate")
+			return ctrl.Result{}, err
+		}
 
-	// your logic here
+	}
 
-	return ctrl.Result{}, nil
+	return ctrl.Result{
+		Requeue:      true,
+		RequeueAfter: time.Minute * 5,
+	}, nil
```

Comment:
requeue after 5 mins, maybe it's too often? we might not really need it that often, once the role is created we wont usually don't need to do anything

Reply:
I think it's okay? Otherwise it will take a long time to create and delete clusters.

Reply:
with kubebuilder, the events like create and delete trigger immediate action, this is just that first reconciliation finishes when it should run again

but leaving 5min for now